DOCS-9144 updated CORS policy to address defects (#329)
mulesoft/docs-api-manager: modules/ROOT/pages/cors-policy.adoc

| No return codes exist for this policy
|===

> 

Reviewer: edit instructions:

Hi Kim, would you please review this PR:
Type: CE
Scope: :green:
Due: EOD Monday, if possible.
Please omit all commented text (//)

CORS is a mechanism by which a web application can access resources that are defined in another domain. Browsers implement this standard by default. The CORS policy complies with the https://fetch.spec.whatwg.org/[CORS W3C recommendation] standards.

> To know why you might need to apply the CORS policy to your backend, you must first understand what are origins and cookies, and how a web page can be misused.

Reviewer: ```suggestion
To know why you might need to apply the CORS policy to your backend, you must first understand _origins_, _cookies_, and how they can be manipulated to violate web page integrity.
```

Author: Done! Thanks!

> == Origin

Reviewer: How about this?
```suggestion
== Introduction to the Web Origin Concept
```

Reviewer: So, I'm sort of expecting an intro. sentence that tells me what an origin is, before telling me what it consists of.

Author: Done! Thanks!

At a very basic level, an origin consists of:

* URI Scheme: `http://``
* Host Name: `www.example.com`

> * Port Number: `8080`

Reviewer: ```suggestion
* URI scheme: `http://``
* Host name: `www.example.com`
* Port number: `8080`
```

Author: Done! Thanks!

> Any two origins are considered equal only if all these three parameters match. For more information, see https://tools.ietf.org/html/rfc6454[RFC 6454 - The Web Origin Concept]

Reviewer: What is the purpose of determining origin equality? How is this part of the introduction to what an origin is and how you use it?

Reviewer: I think we need something more specific here about how origins are part of web page integrity.
Alternatively, I would leave this as-is but limit the following topic to only an overview of cookies, and then have a separate topic about how origins and cookies affect web page integrity.

Author: Done! Thanks!

> == HTTP Cookies

Reviewer: Per previous comment, I would limit this topic to what cookies are, and then make a separate topic of how origins and cookies affect web page integrity.

Reviewer: Consider a future info reorg. in which you have two separate Level-1 topics:

- One titled something about why you want to use CORS to achieve web page security and integrity (origins, cookies)
- One titled something about how to use CORS components (algorithm, request headers, response headers, etc._

Reviewer: Okay, strict copyedit from here onward....

Author: Done! Thanks!

> HTTP Cookies are a reliable mechanism for websites to retain stateful information. Most commonly, web servers use authentication cookies to know whether a user is logged in and using which account are they logged in. For more information, see https://www.ietf.org/rfc/rfc6265.txt[RFC 6265].

Reviewer: ```suggestion
Websites use HTTP cookies to retain stateful information. Most commonly, web servers use cookies to authenticate users, to know whether a user is logged in, and to know which account users are logged in to. For more information, see https://www.ietf.org/rfc/rfc6265.txt[RFC 6265].
```

Author: Done! Thanks!

> To avoid (or deter?) malicious attackers from exploiting cookies when a web page invoked another web page, the SOP (Same-origin Policy) was introduced. For example, attackers might obtain your bank’s login cookie and query the bank’s API on your behalf if you attempted to access a malicious web page.  

Reviewer: ```suggestion
SOP (same-origin policy) deters malicious attackers from exploiting cookies when one web page invokes another. For example, you might log in to your bank's web page using a cookie that an attacker might be able to obtain and exploit to query the bank’s API on your behalf.  
```

Author: Done! Thanks!

> With the SOP, scripts can access data from another web page only if the caller web page has the same origin. For more information about origins, see https://en.wikipedia.org/wiki/Same-origin_policy[Wikipedia].

Reviewer: ```suggestion
With the SOP, scripts can access data from a target web page only if it has the same origin as the caller web page. For more information about origins, see https://en.wikipedia.org/wiki/Same-origin_policy[Wikipedia].
```

Author: Done! Thanks!

> The following example illustrates the origin of the web page the same as the origin section, http://www.example.com:8080:

Reviewer: What following example?
Also, this is an incomplete sentence: "...illustrates the origin of the web page the same as the origin section...."
Finally, there is no period.

Author: Done! Thanks!

image:cors-policy-sop-example.png][SOP Example,95%,85%]

> Because the SOP is highly restrictive, access from one sub-origin to another or to external hyperlinks do not work on webpages. For example, if an origin `www.testapply.com` has two sub-origins, `www.eng.testapply.com` and `www.docs.testapply.com`, communication between the two sub-origins is denied. Additionally, any hyperlinks to external websites from any of the sub-origins is also denied.

Reviewer: What are suborigins and how do external hyperlinks relate to attacks?

Reviewer: GLOBAL: webpages > web pages

Reviewer: sub-origin > suborigin, per our style to avoid hyphens following prefixes in most cases. See style ref.

Author: Done! Thanks!

To circumvent this problem, web browsers implemented the CORS standard, which validates the web server and accepts the request if the validation is successful.

> In the bank example mentioned in the previous section, the bank’s login server can implement the server-side protocol for CORS, allowing only to be queried from their experience API. Once the check is in place, no web page can query the login API from a domain other than the bank’s domain.

Reviewer: Bad grammar, and what does this mean?  "...allowing only to be queried from their experience API"?

Reviewer: Per style ref., GLOBALLY replace "once" with "after" if that is what you really mean (instead of "one time").

Reviewer: "in place" is idiomatic; reword for globalization.

Author: Done! Thanks!

> == CORS on the Client-Side

Reviewer: ```suggestion
== CORS on the Client Side
```

Author: Done! Thanks!

> The CORS algorithm on the client-side consist of the following parts:

Reviewer: Why does this refer to the CORS algorithm when we have not yet defined what that is (next topic)?

Reviewer: Is a "part" of an algorithm an action, such as determining and executing?
These seem like stages, not parts.

Author: Done!

> * Determining whether the request is complex (and potentially dangerous) and sending a ‘preflight’ request to verify whether the server accepts the origin.

Reviewer: Because "preflight" is idiomatic, how about "sending a preliminary _preflight_ request"?

Also, `  `  markup is reserved for syntaxy things; quotation marks are replaced by italic (_ _) if we are introducing a term.

Reviewer: Does verification only occur if the request is complex?

Author: Done! Thanks!

> * Executing the actual request and validating that the server responds correctly and accepts the origin.

Reviewer: So, the two stages or steps or whatever of this algorithm are as simple as "testing" before execution and "validating" afterward.

> A ‘preflight’ is a request (using OPTIONS as HTTP method) for the web browser to inform the backend server the identity (origin and a few other headers) of the web page that is trying to perform the request. If the backend does not accept the origin, the backend server responds to the request without a specific header (Access-Control-Allow-Origin). The client then understands that the page’s origin is not allowed in that server and does not execute the actual request.

Reviewer: ```suggestion
A _preflight_ is a preliminary request (using OPTIONS as the HTTP method) from the web browser to the backend server to test the identity (origin and a few other headers) of the web page that is trying to perform the request. If the backend does not accept the origin, the backend server responds to the request without a specific header (Access-Control-Allow-Origin). The client then understands that the page’s origin is not allowed in that server and does not execute the actual request.
```
Is "the client" in the final sentence being equated to "the web browser" in the first? We have not identified anything yet as a client.

Author: Done! Thanks!

The following diagram shows the XMLHttpRequest (XHR) in JavaScript flow for determining whether to execute the actual request:

> 

Reviewer: Where is the diagram?

Reviewer: "...in JavaScript flow" > "in a JavaScript flow"?

Author: Done! Thanks!

> == CORS Algorithm

Reviewer: Why is this topic separated from the previous, which describes the stages of the algorithm?

Author: Done! Thanks!

The CORS policy is a two-step algorithm: 

> * If the request is considered to be ‘complex’ (see the previous client side diagram for XHR) then a preflight request needs to be executed. 

Reviewer: ```suggestion
* If the request is considered to be complex (see the previous client side diagram for XHR), then a preflight request is executed. 
```

Author: Done! Thanks!

+
If the server does not return the proper CORS response headers for the preflight, the client library (XHR in the previous example) is not allowed to perform the actual request.

> * If the preflight response is correct and complete, then the client library  executes the actual request, which includes certain CORS headers. 

Reviewer: ```suggestion
* If the preflight response is correct and complete, then the client library executes the actual request, which includes certain CORS headers. 
```

Author: Done! Thanks!

+
The client library then validates the CORS headers of the response. If some of the required headers are missing, then again the client library is obliged to block the response from reaching the client, which is usually a web page.

> 

Reviewer: This subtopic is largely repetitive of the previous. This topic set needs reorganization at some point.

Reviewer: Over the course of these texts, we have gone from "web page" to "client." Please clarify at the beginning of this topic set. Avoid using more than one term for one thing; it confuses the audience.

Author: Done! Thanks!

> The following section details the headers for the request and response for the CORS algorithm.

Reviewer: Per our style reference and GLOBALLY, do not refer to "following" or "previous" text "sections" or "topics."
Each subtopic must be independent and reusable.
This sentence could easily be modified to serve as the introduction to the following subtopic.

Author: Done! Thanks!

== CORS Request Headers

* Origin: The origin making the cross origin request.
* Access-Control-Request-Method:  The method that is invoked in the actual request . This header is sent in the preflight request.

> * Access-Control-Request-Headers: Custom headers that are sent in the actual request. This header is sent in the preflight request.

Reviewer: If you have more than one phrase or short sentence following the list item, you may not use the colon format: see the style guide entry for bullet lists. 
```suggestion
* Origin
+
The origin making the cross-origin request.
* Access-Control-Request-Method
+
The method that is invoked in the actual request. This header is sent in the preflight request.
* Access-Control-Request-Headers
+
Custom headers that are sent in the actual request. This header is sent in the preflight request.
```

Author: Done! Thanks!

> [NOTE]

Reviewer: See style ref. regarding use of NOTE

Author: Done! Thanks!

> For the GET or HEAD methods, the standard defines a specific list of headers that are considered as ‘simple’ and that do not need to perform a preflight. For headers considered as ‘custom’  by the standard, a preflight is executed for GET and HEAD requests. See the example of XHR to validate the path in which the client does not have to perform a preflight.

Reviewer: ```suggestion
For the GET or HEAD methods, the standard defines a specific list of headers that are considered simple enough to not require a preflight. For custom headers, a preflight is required (see the previous XHR example).
```
"custom" seems to be used here where "complex" was used earlier: standardize terminology.

Shouldn't this information have been given to the audience earlier, when "complex" was first mentioned in regard to a preflight?

Author: Done! Thanks!

== CORS Response Headers
	

> The headers included in a response differs based on whether the request is a preflight or an actual request:

Reviewer: ```suggestion
The headers included in a response depend on whether the request is a preflight or an actual request:
```

Author: Done! Thanks!

* Access-Control-Allow-Origin: Obligatory in every response.
+
If this header is not present in the response, the browser or client library blocks the response from reaching the web page. The wildcard “*” can be used to represent any origin.
* Access-Control-Allow-Methods: Allowed methods that can be executed.
+

> This headers is returned in the OPTIONS request (preflight). The server might respond with a list of allowed methods, delegating the validation task of whether its actual request is supported to the client.

Reviewer: ```suggestion
This header is returned in the OPTIONS request (preflight). The server might respond with a list of allowed methods, delegating the validation task to the client.
```

Author: Done! Thanks!

* Access-Control-Allow-Headers: Allowed headers in the actual request.
+ 

> This header works analogous to Access-Control-Allow-Methods.

Reviewer: ```suggestion
This header works in a manner similar to how Access-Control-Allow-Methods works.
```

Author: Done! Thanks!

> * Access-Control-Allow-Credentials: Notifies the client, with a Boolean value, whether the actual request can be made with a cookie.

Reviewer: This is beginning to stretch the idea of a short text that can follow a bullet-list item and colon....

Author: Done! Thanks!

> * Access-Control-Expose-Headers: Notifies the browser or client library the list of  headers that can be accessed by  the web page that executed the request. 

Reviewer: ```suggestion
* Access-Control-Expose-Headers: Provides the browser or client library with a list of headers that can be accessed by the web page that executed the request. 
```

Author: Done! Thanks!

+

> The HTTP library doing the CORS request, will only show the expose headers to the web page. This allows further filtering of the headers for privacy.

Reviewer: ```suggestion
The HTTP library responding to the CORS request exposes only the headers to the web page, providing further privacy and security.
```

Author: Done! Thanks!

> * Access-Control-Max-Age: Notifies the client the number of seconds during which the browser can avoid performing a second preflight for the same request. 

Reviewer: ```suggestion
* Access-Control-Max-Age: Specifies the number of seconds before the browser performs a second preflight on a request. 
```

Reviewer: BREAK to commit to avoid Github failure

Reviewer: RESUME

Author: Done! Thanks!

> == Public resource

Reviewer: ```suggestion
== Public Resource
```

Author: Done! Thanks!

> If you need to bypass the SOP of the browser, MuleSoft provides you the option of configuring a public resource that enables the API gateway policy to mirror the preflight data in the response. This also ensures that the actual request is correctly updated with all the CORS headers, such that the browser accepts the response. 

Reviewer: ```suggestion
If you need to bypass the browser SOP, configure a public resource that enables the API gateway policy to mirror the preflight data in the response. This ensures that the actual request is correctly updated with all the CORS headers so that the browser accepts the response. 
```

Author: Done! Thanks!

//Mariano--is it sin

== Ordering

> By design, the CORS policy is always applied first in order by API gateway and you cannot change this order of execution. If a protected request using OPTIONS is sent to an application that has the CORS policy applied, the request does not reach the protected resource. According to the https://fetch.spec.whatwg.org/#resource-preflight-requests[CORS specification], all OPTIONS requests are considered preflight.

Reviewer: ```suggestion
The CORS policy is always applied by the API gateway first. If a protected request using OPTIONS is sent to an application that has the CORS policy applied, the request does not reach the protected resource. According to the https://fetch.spec.whatwg.org/#resource-preflight-requests[CORS specification], all OPTIONS requests are considered preflight.
```
"first" before what other executables?
And why does this mean that an OPTIONS request fails? because it doesn't pass the test? why?

Author: Done! Thanks!

> == Groups

Reviewer: I suggest combining this with Public Resource, since the first sentence refers to it.

Author: Done! Thanks!

> If the public resource option is not a strict enough configuration for your environment, you can define multiple groups for the different origins that query your API. Each group applies to a list of origins and can specify different methods, headers, preflight caching time, and expose headers.

Reviewer: ```suggestion
If the public resource option is not secure enough for your environment, define multiple groups for the different origins that query your API. Each group applies to a list of origins and can specify different methods, headers, preflight caching time, and expose headers.
```
Is "expose" a type of header or are you using this as a verb? If so, the proper wording is...:

Each group applies to a list of origins and can specify different methods, headers, and preflight caching time, and can expose headers.
```

Author: Done! Thanks!

> == Wildcards on configuration

Reviewer: ```suggestion
== Configuration Wildcards 
```

Author: Done! Thanks!

> Wildcards are accepted in the Origins and Headers sections for a group configuration. Use this configuration, for example, when you need a public resource but restricting the accepted HTTP methods. For both sections, an asterisk “*” is the wildcard. 

Reviewer: Best guess:
```suggestion
Wildcards (`*`) are accepted in the Origins and Headers sections of a group configuration. Use wildcards when you need a public resource but must restrict the accepted HTTP methods. 
```
In the original "Use this configuration," are you referring to wildcards or to a group config?

Author: Done! Thanks!

> If you have configured multiple groups and one of the groups uses a wildcard origin, the none-wildcard ones override the wildcard configuration for the specific origins defined in those groups.

Reviewer: Best guess:
```suggestion
If you have a mixture of groups that use wildcard origins and groups that do not, the nonwildcard settings override the wildcard configurations.
```

Author: Done! Thanks!

> == CORS does not seem to be applied

Reviewer: Is this supposed to be a troubleshooting topic?
Reword title to be a peer of the others.

Author: Done! Thanks!

> If you use tools, such as curl or Postman to test the CORS policy, neither does the tool add the CORS request headers, nor does it perform the preemptive preflight request in case of a complex request.

Reviewer: ```suggestion
If you use tools such as curl or Postman to test the CORS policy for a complex request, the CORS request headers are not added and the preflight does not occur.
```

Author: Done! Thanks!

> If no CORS headers are sent or improper headers are used, API gateway CORS policy does not add any CORS response headers and it appears as the policy is not applied. Ensure that you use the libraries that implement the CORS protocol for you, or review the specification to ensure that you are configuring the correct headers.

Reviewer: Best guess:
```suggestion
If no CORS headers are sent or improper headers are used in the request, the API gateway CORS policy does not add any CORS response headers, giving the impression that the policy is not applied. Ensure that you use the libraries that implement the CORS protocol specific to your environment, or review the specification to ensure that you are configuring the correct headers.
```

Author: Thanks! Priya!

//image::cors-policy.png[]

> == Configuring CORS for non-API Gateway Mule Environments

Reviewer: ```suggestion
== Configuration of CORS for non-API Gateway Mule Environments
```

Author: Done! Thanks!

The CORS algorithm works on the web server and on the client-side for the web page that requested the information. The client-side algorithm in the CORS policy is implemented by:

* Determining whether the request is complex (and potentially dangerous) and sending a preliminary _preflight_ request to verify whether the server accepts the origin.
* Executing the actual request and validating that the server responds correctly and accepts the origin.

A _preflight_ is a preliminary request (using OPTIONS as the HTTP method) from the web browser to the backend server to test the identity (origin and a few other headers) of the web page that is trying to perform the request. 

If the backend does not accept the origin, the backend server responds to the request without a specific header (Access-Control-Allow-Origin). The client then understands that the page’s origin is not allowed in that server and does not execute the actual request.

The following diagram shows the XMLHttpRequest (XHR) in a JavaScript flow for determining whether to execute the actual request:

image::cors-policy-xhr-diagram.png[]

> When the request is executed and it reaches the backend, the server side of the algorithm is initiated in two steps:

Reviewer 2: This is actually a summary of the interaction between backend and browser.
Change this for smth like "The detail interaction between the browser and the server goes as follows:"

Author: done! Thanks!

| Support credentials | Whether the policy supports credentials, such as cookies, authorization headers, and TLS client certificates | No
|===

== FAQs

*The CORS policy does not seem to be applied.*

If you use tools such as curl or Postman to test the CORS policy for a complex request, the CORS request headers are not added and the preflight does not occur.

If no CORS headers are sent or improper headers are used in the request, the API gateway CORS policy does not add any CORS response headers, giving the impression that the policy is not applied. Ensure that you use the libraries that implement the CORS protocol specific to your environment, or review the specification to ensure that you are configuring the correct headers.

> //image::cors-policy.png[]

Reviewer 2: need to remove this comment?

Author: Done! Thanks!
